### ISSUE TITLE
Create wrapper for Jikan API v4 beta

### ISSUE THREAD

**Issue by abhinavk99** [abhinavk99/jikanpy#94]
2 new classes would be created, Jikan4 and AioJikan4 that could be imported like `from jikanpy import Jikan4`

Endpoint: https://api.jikan.moe/v4
Docs: https://docs.api.jikan.moe/

Once Jikan API v3 gets deprecated, Jikan4 will become the main class and Jikan will become Jikan3 as seen below.
```
Jikan => Jikan3
AioJikan => AioJikan3
Jikan4 => Jikan
AioJikan4 => AioJikan
```

**Comment by seanbreckenridge** [comment]
Sounds good

Not sure if there are OpenAPI consumers in python/if you wanted to use that - v4 would provide a schema file which could be converted into structs/request interfaces. Not sure how that affects documentation etc.

I can look into that/can discuss on discord if that's a possibility

**Comment by abhinavk99** [comment]
In Nekomata's message he said not to use the tools yet since v4 is in beta, but we could look into it. It looks like the generator at https://github.com/OpenAPITools/openapi-generator can be combined with ReDoc or documentation generators.

**Comment by GhostLyrics** [comment]
Yo. I guess it's time to revisit this [based on this document that's now part of Jikan's responses](https://docs.google.com/document/d/172RQ9wWiXqOnGqjXrV3cxMNceiqwCjxjprSFuyLwQJM/). 

Quote from linked doc:
> Timeline
> 1️⃣ 1st March 2022 - Deprecation
> The v3 API will enter into a deprecation state from 1st March 2022. 
> 👉 You will be receiving a warning through headers and in the JSON body as mentioned above.
>
> 2️⃣ 1st May 2022 - Maintenance Ends
> The v3 API will no longer receive any maintenance from 1st May 2022. 
> 👉 This includes parser fixes.
>
> 3️⃣ 1st September 2022 - Brownout
> The v3 API will enter into a brownout state from 1st September 2022.
> 👉 v3 Endpoints will return HTTP 410 GONE for the first 15 minutes of every other hour to alert anyone still using v3.
>
> 4️⃣ 1st October 2022 - Sunsetting
> The v3 API will be discontinued from 1st October 2022.
> 👉 v3 Endpoints will no longer be available on the public Jikan REST API.

**Comment by abhinavk99** [comment]
To be perfectly honest, I'm not sure when I'll be able to get to this as I've gotten a job and am now more busy. It'd be great if you or someone else could take this up and I can review the changes. I think we probably just need to use the OpenAPI wrapper generator and update the docs.

**Comment by GhostLyrics** [comment]
I'm sorry, I have neither the time nor enough interest to maintain this library. 😓

**Comment by slashbeast** [comment]
I am interested in v4 support so I might produce some results. No idea how much v4 interface is different from v3 though. Did either of you did some research in this area?

**Comment by seanbreckenridge** [comment]
Its pretty different -- you may be able to leverage the openapi file which describes the schema though:

see docs

https://docs.api.jikan.moe/

dont know how good this code is, or if we want to use `attrs` and `httpx`, but its at least a starting point and something to reference. it could be great, idk:

```
pip install pipx
pipx install openapi-python-client --include-deps
pipx run openapi-python-client generate --url 'https://raw.githubusercontent.com/jikan-me/jikan-rest/master/storage/api-docs/api-docs.json'
```

generates generates a bunch of models and an API structure:

![image](https://user-images.githubusercontent.com/7804791/158082824-b060fea0-0948-46aa-81ee-d425c514ee1d.png)

**Comment by seanbreckenridge** [comment]
would probably move setup.py/v3 to a separate subdir, the generated code uses poetry so that would be in `v4`

could create a buildscript to create the wheels so they could be pushed to pypi separately

will take a shot at it this weekend if no one else is already doing something

**Comment by seanbreckenridge** [comment]
Been working on a basic restructure, see #99

**Comment by lizelive** [comment]
look forward to this

**Comment by seanbreckenridge** [comment]
To just copy paste what was said in `#wrapper-devs`, regarding #99 

There's incompatibility in types with the schema and the generator is super strict so it doesn't really work. 

Edit: actually, part of it did get fixed [here](https://github.com/jikan-me/jikan-rest/issues/245), but all the other requests still need to be checked

If the schema did match perfectly would probably have been easier but I dont think either of us are as free to create an entire wrapper from scratch right now

**Comment by nerg4l** [comment]
I try to maintain the OpenAPI Schema of Jikan. I use it in a project to generate a python client and experienced the same problem with the generated code. I do not know at the moment how could we validate the schema automatically in Jikan but if you have any proposal, feel free to reach out.

**Comment by MahdeenSky** [comment]
Looks like it has been discontinued now.
```jikanpy.exceptions.APIException: HTTP 503 - status=503, type=BadResponseException, message=v3 has been discontinued. For more information visit https://bit.ly/jikan-v3-deprecation, upgrade=Please upgrade to v4: https://docs.api.jikan.moe/ for id=48675, endpoint=anime```

**Comment by abhinavk99** [comment]
Fixed in #101
